pennylane/resources/first_quantization.py

```diff
@@ -0,0 +1,221 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for estimating the number of logical qubits and
+non-Clifford gates for quantum algorithms in first quantization using a plane-wave basis.
+"""
+# pylint: disable= too-many-arguments
+from functools import partial
+from scipy.optimize import minimize_scalar
+from pennylane import numpy as np
+
+
+def _cost_qrom(k, lz):
+    r"""Return the number of Toffoli gates needed for erasing the output of a QROM.
+    ￼
+    The expression for computing the cost is taken from
+    [`arXiv:2105.12767 <https://arxiv.org/abs/2105.12767>`_].
+
+    Args:
+        k (int): parameter taken to be a power of 2
+        lz (int): sum of the atomic numbers of nuclei
+
+    Returns:
+        int: the cost of erasing the output of a QROM
+
+    **Example**
+
+    >>> _cost_qrom(4, 100)
+    23
+    """
+    cost = 2**k + np.ceil(2 ** (-k) * lz)
+
+    return int(cost)
+
+
+def _cost_qrom_min(lz):
+    r"""Return the minimum number of Toffoli gates needed for erasing the output of a QROM.
+    ￼
+    The expression for computing the cost is taken from
+    [`arXiv:2105.12767 <https://arxiv.org/abs/2105.12767>`_].
+
+    Args:
+        lz (int): sum of the atomic numbers of nuclei
+
+    Returns:
+        int: the minimum cost of erasing the output of a QROM
+
+    **Example**
+
+    >>> _cost_qrom_min(100)
+    21
+    """
+    n = minimize_scalar(partial(_cost_qrom, lz=lz)).x
+
+    cost = min(_cost_qrom(np.floor(n), lz), _cost_qrom(np.ceil(n), lz))
+
+    return cost
+
+
+def unitary_cost(n, eta, omega, error, lamb, br=7, charge=0):
+    r"""Return the number of Toffoli gates needed to implement the qubitization unitary operator.
+    ￼
+    The expression for computing the cost is taken from
+    [`arXiv:2105.12767 <https://arxiv.org/abs/2105.12767>`_].
+
+    Args:
+        n (int): number of basis states
+        eta (int): number of electrons
+        omega (float): unit cell volume
+        error (float): target error in the algorithm
+        lamb (float): 1-norm of the Hamiltonian
+        br (int): number of bits for ancilla qubit rotation
+        charge (int): total electric charge of the system
+
+    Returns:
+        int: the number of Toffoli gates needed to implement the qubitization unitary operator
+
+    **Example**
+
+    >>> n = 100000
+    >>> eta = 156
+    >>> omega = 169.69608
+    >>> error = 0.01
+    >>> lamb = 5128920.595980267
+    >>> unitary_cost(n, eta, omega, error, lamb)
+    12819
+    """
+    l_z = eta + charge
+    l_nu = 4 * np.pi * n ** (1 / 3)
+
+    n_eta = np.ceil(np.log2(eta))
+    n_etaz = np.ceil(np.log2(eta + 2 * l_z))
+
+    n_p = np.ceil(np.log2(n ** (1 / 3) + 1))
+    n_t = np.log2(np.pi * lamb / (0.01 * error))
+    n_r = np.log2((eta * l_z * l_nu) / (0.01 * error * omega ** (1 / 3)))
+    n_m = np.log2(
```

Comment:
Same issues ("+ replaced with *", etc.) as in the other commit #2720 for the definition of `n_m`